openstates/importers/base.py

```diff
@@ -169,15 +169,28 @@ def resolve_bill(self, bill_id: str, *, date: str) -> typing.Optional[_ID]:
         if bill_transform_func:
             bill_id = bill_transform_func(bill_id)
 
-        # move the start_date up a bit in case the event is on the last day of a session to compare with end_date
+        # Some steps here to first find the session that match the incoming event using the event date
```

Comment:
You make a good point about increasing queries: in the past lots of queries in resolve_bill() has caused some real DB issues. So I think it could indeed be worth a bit more refactor to move this logic somewhere else. 

The Importer base class has a property already called `session_cache` ([see here](https://github.com/openstates/openstates-core/blob/bde7c3ffc2256f2e1be53fe7f7b66745c712fe07/openstates/importers/base.py#L130)). Here's what I'm thinking:

- We could add a new method to base importer `get_all_sessions()` but there is a possible bad scenario where `get_session()` (which already exists and is used with Bills and Votes) loads a single session or two into `session_cache`. That creates a situation where it is impossible to know if we have *all* of the sessions or not. So we don't want to set up a scenario where `get_all_sessions()` logic is set up to fail by `get_session()`
- So then maybe make a change to `get_session()` ([here](https://github.com/openstates/openstates-core/blob/bde7c3ffc2256f2e1be53fe7f7b66745c712fe07/openstates/importers/base.py#L142)) so that it loads *all* sessions into `session_cache` instead of just the one requested session (but it would still return the requested session). Basically it is not any less performant to query for 16 sessions than it is to query for one session, so why not load them all anytime some client code wants a session loaded from the DB? 
- `resolve_bill()` calls `get_session()` or some new method `get_all_sessions()`, meaning that it is going to get back all the sessions as a list of dicts, and there will be only one DB query to get them.
- then instead the logic here loops through the list of sessions to make comparisons and find the right one (instead of db query with WHERE clauses).

OR - leave the logic here but implement a new `self.current_session_id_cache` property to save it for use by subsequent events.

All that said - there are generally many fewer events returned by an events scraper (vs. bills/votes). And I think only Event importer uses `resolve_bill()`. So we definitely could just merge this in and wait to see if performance issues appear before making these kind of changes.

Reply:
Very good point, thank you for the additional insight. I will take a good.